### PR TITLE
Add support for monadic let operators (fixes #1868)

### LINKED ISSUE #1868: Support let operators à la OCaml
tl;dr see https://github.com/ocaml/ocaml/pull/1947

We could support them in F* syntax to have a more flexible do notation, where one can combine different binds as needed (@beurdouche recently asked for this) and to make extracted programs more readable, especially for DM4F/DM4A.

### PR DESCRIPTION
**Pull request by W95Psp**
Hi,

This PR adds support for monadic let operators (see #1868), so that one can write monadic things like:

```fstar
let (let?) (x: option 'a) (f: 'a -> option 'b): option 'b
  = match x with
  | Some x -> f x
  | None   -> None

let (and?) (x: option 'a) (y: option 'b): option ('a * 'b)
  = match x, y with
  | Some x, Some y -> Some (x, y)
  | _ -> None

let foo (x: option string)
    	(f: string -> option (list (option (int * int))))
	: option int
  = let? x in
    match? f x with
    | []   -> Some 1
    | hd::_ -> let? a, b = hd in
      	     Some (a + b)
```

The PR adds three new tokens in the lexer (for operator `let`, `and` and `match`), and a new constructor in the surface AST (`LetOperator`).
I also added `let?` and `and?` in `FStar.Option`, it feels like it would be useful to have them around in `ulib`.

### CODE REVIEW

**Comment by nikswamy** [comment]
This looks really cool, though I have a few questions:

How does this compare to F*'s existing monadic notation, `x <-- e ; e'` and `e ;; e'`? E.g., it seems that `let?` supports any pattern on the LHS, rather than just a variable?

Do they interact in any way? 

Does this new notation play well with typeclasses? I guess I can define `let?` in a monad polymorphic way and this would work?

How is `let?` desugared?

Is `let?` the right name for it? E.g., this library in OCaml has similar stuff, named differently: https://github.com/janestreet/ppx_let

**Comment by W95Psp** [comment]
Thanks for the feedback!

Concerning the naming, I actually followed more or less what OCaml does (https://v2.ocaml.org/manual/bindingops.html), that is, the lexer recognizes any `let`, `and` or `match` followed by one or more character in the class `!$%&*+-./<=?^|~:`. (and I named the monadic let for options `let?` following Rust operator `?`)

A let binding of the shape `let* x = foo and y = bar in body x y` is desugared into `let_op_Star (and_op_Star foo bar) body`.

There's no conflict in the grammar for monadic lets to accept patterns in their LHS, so yes, that's an improvement over the do notation. Also, you can have multiple different let operators in one same scope (e.g. `let?` for options then `let::` for lists) which I find is quite nice for common things such as options or eithers. And the last nice thing is the support for monadic `match`es as well.
Those let operators should play nicely with typeclasses, but I havn't tried that yet.

File: src/parser/FStar.Parser.AST.fst [hunk 1]
```diff
@@ -516,6 +518,59 @@ let compile_op arity s r =
...
+                                          | None -> None
+                                          | Some acc ->
+                                              match x with
+                                              | Some (op, _) -> Some (acc ^ op)
+                                              | None -> None)
+                            (Some "")
+                            (List.map name_of_op s)
+           in
+           map_opt maybeop (fun o -> (o, None))
+    else None
```
Comment (nikswamy) [inline src/parser/FStar.Parser.AST.fst:572]: This seems to be just a move of string_to_op from Resugar ... which some formatting changes in the default case, right?

**Comment by nikswamy** [comment]
This looks really great and useful ... and I'm keen to use this syntax right away!

The code looks nice and clean too and it's great that all the changes are scoped to the frontend syntax.

I'm going to merge this now, however it would be great if you could add just a little more documentation:

For example, in CHANGES.md, could you please point to the nice sample file you have in examples/misc/MonadicLetBindings.fst?

And in MonadicLetBindings.fst, it would be great to have a comment describing that 

```
let? x = ex
and? y = ey
and? z = ez in
body
```

desugars to

```
let? (and? (and? ex ey) ez) (fun r ->
  let (x, y), z = r in
  body
```


Also, FYI, the existing do notation is documented here: http://fstar-lang.org/tutorial/book/part2/part2_par.html#a-first-taste-the-state-monad ... this isn't particularly easy to discover. If you have any thoughts as to where we could better document this monadic let/and/match notation, that would be great.

Thanks!

**Comment by nikswamy** [comment]
Also, what do you think about adding similar support for sequencing, e.g., 

```
e1 ;? e2
```

desugars to 

```
let? _ = e1 in e2
```

**Comment by W95Psp** [comment]
Thanks for merging and for the nice comments! :)

I've added some desugaring examples, the emacs-mode tip, some
pointers in CHANGE.md, and more explanations in PR#2666 then!

## Pretty printing issues
The printing is indeed not very nice, I've addressed it only partially
in PR#2666. Consider the following let bindings:

```fstar
let ( let* )       foo  = 0
let    ( * )       foo  = 0
let ( let* ) = fun foo -> 0
let (    * ) = fun foo -> 0
```

Printing the two latter preserve the syntax, but the two formers are
printed with their `op_…` form:

```fstar
let op_let_Star    foo  = 0
let     op_Star    foo  = 0
let ( let* ) = fun foo -> 0
let ( * )    = fun foo -> 0
```

This bad printing is due to the fact the parser rule
[`letbinding`](https://github.com/FStarLang/FStar/blob/f327b8a6bcdd25d09d3c241562cbd404fc111a9a/src/parser/parse.mly#L324)
when parsing a let binding with binders uses the rule
[`lidentOrOperator`](https://github.com/FStarLang/FStar/blob/f327b8a6bcdd25d09d3c241562cbd404fc111a9a/src/parser/parse.mly#L641)
that transforms an operator (eg `*`) into its explicit name (eg
`op_Star`). The parser parses `let ( * ) = …` and `let op_Star = …`
into the exact same surface AST.

I guess we should just split parser rules involving `letbinding` to
produce `PatOp` instead of `PatVar`, but this will be an issue for
`val (+) = …`.

## Sequences
I'm totally in favor of support for sequence operators!

Though, I think that allowing `x ;* y` but not `x <-- ;* y` would be a
bit odd from the user's point of view. The latter doesn't cause any
conflict. I've implemented some support for those notations in PR#2666
as well.

File: src/parser/FStar.Parser.AST.fst (hunk 1 above)
Comment (W95Psp) [reply src/parser/FStar.Parser.AST.fst:572]: Yes, the formatting changes are side effects of some experiments I did earlier, I shouldn't have included those changes in the commit!

**Pull request merged**